Pull sample rate from DAS table rather than PH5toMSeed.get_nonrestric…

timronan opened the pull request:
…ted_segments.

Thank your for contributing to PH5

Before submitting a PR, please review the pull request guidelines:
https://github.com/PIC-IRIS/PH5/blob/master/CONTRIBUTING.md#submitting-a-pull-request

### What does this PR do?
Pulls sample rate directly from DAS table for command ph5toms. This allows ph5 to ms to run on channels that have a sample rate that is different than 100 Hz. 

### Relevant Issues?
#424

### Checklist
- [x] This PR is not directly related to an existing issue (which has no PR yet).
- [x] All tests pass.
- [x] Any new or changed features have are documented.  **No new features. Just fixing old ones. **
- [x] Changes have been added to `CHANGELOG.txt` .
- [x] First time contributors have added your name to `CONTRIBUTORS.txt`

timronan commented:
This is ready for review. Please test and let me know what you think.

dsentinel commented:
Is this the route to take, or should we just include a validation that makes sure all SR exist and are the same?

I'm unsure about the repercussions of this across other tools, i.e. does ph5tosegy need to be adjusted as well?
@damhuonglan can you confirm that this works as expected, and coordinate with the datagroup for the same.

timronan commented:
I already have a validation check included in this PR. Even though this issue originates in the data, the code is not optimally factored and is currently using a sample rate that does not always reflect the sample rate of the data. More refactoring may need to occur so the DAS table sample rates are used for all PH5 functions. The problem with not fixing this issue in the code, is that we already have data which contain mismatching sample rates between the DAS and Array tables. We may also output incorrect information if this code is not updated. It would be better if these data were made available.

damhuonglan commented on ph5/clients/tests/test_ph5availability.py:
```diff
@@ -1120,8 +1120,8 @@ def test_main(self):
                     "                      latest\n"
-                    "AA 0407  -- HHN   2018-12-17T22:20:30.917000Z"
-                    " 2018-12-17T22:20:40.922000Z\n"
+                    "AA 0407  -- HHN   2018-12-17T22:20:30.681998Z"
+                    " 2018-12-17T22:20:40.691998Z\n"
```
Giving -S option should give the same output as original test.

timronan replied:
The test suite was reverted back to the original test and now passes.

damhuonglan commented on ph5/clients/ph5availability.py:
```diff
@@ -473,22 +491,48 @@ def get_availability(self, station='*', location='*',
+                            chan=ph5_channum,
+                            start_epoch=ph5_start_epoch,
+                            stop_epoch=ph5_stop_epoch)
+                        for das in Das_t:
+                            # Sample rates must first be compared than
+                            ph5_sample_rate = das['sample_rate_i']
+                            times = self.ph5.get_availability(
+                                ph5_das, ph5_sample_rate, ph5_channum,
+                                starttime, endtime)
+                            if das['sample_rate_i'] == st['sample_rate_i']:
+                                samplerate_return = das['sample_rate_i']
+                            if times is None:
+                                continue
```
This loop mess up sample rate at the result. It's ok if we don't include sample rate, but when we include sample rate, it gives the wrong result. Please look at station 0407 in array 2 and 3 and compare with das 5553 sample rate 100 and 200.

timronan replied:
The loop was slightly changed to make sure the correct sample rate is used when querying for cut times. 

``` 
                       for das in Das_t:
                            # Sample rates must first be compared than
                            if das['sample_rate_i'] == st['sample_rate_i']:
                                samplerate_return = das['sample_rate_i']
                                ph5_sample_rate = das['sample_rate_i']
                                times = self.ph5.get_availability(
                                        ph5_das, ph5_sample_rate, ph5_channum,
                                        starttime, endtime)
                            if times is None:
                                continue
```

damhuonglan replied:
It doesn't change sample rate but change time for sample rate. It might be ok depend on the condition for you to filter later. I don't have time to read carefully the logic later. The original test were correct for sample rate included. It your code pass that test, it should be ok.

timronan replied:
The original test were correct, it was a good idea to look at the data using HDFview. It allowed me to see my confusion. All of the original test now pass. If you think this is the correct logic we maybe need to add this check into PH5toSXML as well. Currently, PH5toSXML can output the incorrect sample rate.

damhuonglan commented on ph5/clients/ph5availability.py:
```diff
@@ -384,27 +384,45 @@ def get_availability_extent(self, station='*', location='*',
+                            chan=ph5_channum,
+                            start_epoch=ph5_start_epoch,
+                            stop_epoch=ph5_stop_epoch)
+                        for das in Das_t:
+                            ph5_sample_rate = das['sample_rate_i']
+                            earliest, latest = self.ph5.get_extent(
+                                ph5_das, ph5_channum, ph5_sample_rate,
+                                starttime, endtime)
+                            if das['sample_rate_i'] == st['sample_rate_i']:
+                                samplerate_return = das['sample_rate_i']
+                            if earliest is None or latest is None:
+                                continue
```
This loop mess up sample rate at the result. It's ok if we don't include sample rate, but when we include sample rate, it gives the wrong result. Please look at station 0407 in array 2 and 3 and compare with das 5553 sample rate 100 and 200.

timronan replied:
The loop was slightly changed to make sure the correct sample rate is used when querying for cut times. 

``` 
                       for das in Das_t:
                            # Sample rates must first be compared than
                            if das['sample_rate_i'] == st['sample_rate_i']:
                                samplerate_return = das['sample_rate_i']
                                ph5_sample_rate = das['sample_rate_i']
                                times = self.ph5.get_availability(
                                        ph5_das, ph5_sample_rate, ph5_channum,
                                        starttime, endtime)
                            if times is None:
                                continue
```

hrotman-pic commented:
Does this PR also require updates to ph5tostationxml?  I thought that it's currently using the array tables to get the sample rate and other metadata.  Concerned it might lead to undescribed data if ph5toms pulls SR from DAS and ph5tostationxml pulls from array.

timronan commented:
We do need to update ph5tostationxml as well. If ph5tostationxml uses the sample rate from the array table, the output stationxml files could be reporting incorrect metadata information.

timronan commented on ph5/clients/tests/test_ph5availability.py:
```diff
@@ -501,9 +501,9 @@ def test_get_availability(self):
+            ('0407', '', 'HHN', 1545085230.681998, 1545085240.691998), ret, 2))
         self.assertTrue(checkTupleAlmostEqualIn(
-            ('0407', '', 'LHN', 1545085230.681998, 1545085240.69), ret, 2))
+            ('0407', '', 'LHN', 1545085230.681998, 1545085240.691998), ret, 2))
```
The test suite was reverted back to the original test and now passes.

timronan commented on ph5/clients/tests/test_ph5availability.py:
```diff
@@ -1120,8 +1120,8 @@ def test_main(self):
-                    "AA 0407  -- HHN   2018-12-17T22:20:30.917000Z"
-                    " 2018-12-17T22:20:40.922000Z\n"
+                    "AA 0407  -- HHN   2018-12-17T22:20:30.681998Z"
+                    " 2018-12-17T22:20:40.691998Z\n"
```
The test suite was reverted back to the original test and now passes.

timronan commented on ph5/clients/tests/test_ph5availability.py:
```diff
@@ -415,9 +415,9 @@ def test_get_availability_extent(self):
+            ('0407', '', 'HHN', 1545085230.681998, 1545085240.691998), ret, 2))
         self.assertTrue(checkTupleAlmostEqualIn(
-            ('0407', '', 'LHN', 1545085230.681998, 1545085240.69), ret, 2))
+            ('0407', '', 'LHN', 1545085230.681998, 1545085240.691998), ret, 2))
```
The test suite was reverted back to the original test and now passes.

timronan commented on ph5/test_data/metadata/extent_full.txt:
```diff
@@ -2,7 +2,7 @@
 AA 500   -- DP2         500.0 2017-08-09T16:00:00.380000Z 2017-08-09T16:01:00.380000Z
 AA 500   -- DPZ         500.0 2017-08-09T16:00:00.380000Z 2017-08-09T16:01:00.380000Z
-AA 0407  -- HHN         200.0 2018-12-17T22:20:30.917000Z 2018-12-17T22:20:40.922000Z
+AA 0407  -- HHN         200.0 2018-12-17T22:20:30.681998Z 2018-12-17T22:20:40.691998Z
```
The test suite was reverted back to the original test and now passes.

timronan commented on ph5/utilities/ph5validate.py:
```diff
@@ -663,6 +664,20 @@ def check_array_t(self):
                                     continue
                                 else:
                                     track_repeated.append(item)
+                            # Pull the das table using the array table
+                            Das_t = self.ph5.query_das_t(
+                                ph5_das,
+                                chan=channel_id,
+                                start_epoch=station['deploy_time/epoch_l'],
+                                stop_epoch=station['pickup_time/epoch_l'])
+                            for das in Das_t:
+                                if(station['sample_rate_i'] !=
+                                   das['sample_rate_i']):
+                                    error.append('Array sample rate '
+                                                 + str(
+                                                    station['sample_rate_i'])
+                                                 + ' != Das sample rate '
+                                                 + str(das['sample_rate_i']))
```
This is a good point. We maybe need to take the ph5 validate portion out of this pull request. It seems like I was wrong earlier that the array table and DAS table have different sample rates. The problem it was that the sample rate from the wrong array table was being used for ph5availabilty and ph5toms. There doesn't seem to be foreign key that connects that the Array and DAS tables so it's hard to know if we are comparing the correct tables.

timronan commented:
Has there been a chance to do any additional testing on this pull request?

timronan commented on ph5/clients/ph5toms.py:
```diff
@@ -511,32 +510,28 @@ def create_trace(self, station_to_cut, mp=False):
+            if Das_tf['sample_rate_i'] > 0:
+                actual_sample_rate = \
+                    float(Das_tf['sample_rate_i']) / \
+                    float(Das_tf['sample_rate_multiplier_i'])
```
Commit 2565318b06e6e29244d0b561cc2fb801c4b01a61 addresses:
Removing the PH5 validate portion from the pull request. 

Adds a check between the DAS sample rate and Array sample rate as uses the array table sample rates as the actual sample rate:

```
            nt = stc.notimecorrect
            if Das_tf['sample_rate_i'] == stc.sample_rate:
                if Das_tf['sample_rate_i'] > 0:
                    actual_sample_rate = float(
                        stc.sample_rate) / float(stc.sample_rate_multiplier)
                else:
                    actual_sample_rate = 0
            else:
                continue
```
An additional test was also added with to ph5toms_test. 

> For ph5_availability, we need some test that really show the changes work. Since those functions always consider SEED channel, we need some das data that have the same SEED channels but different sample rate for testing. 

I may be missing something but it doesn't seem like the das tables have a field for SEED channel, SEED station or SEED Location. That is part of the challenge with coding this patch, there does not seem to be a robust foreign key that links the DAS and Array tables. 

DAS Example:
```
{'time/micro_seconds_i': 380000, 'sample_rate_i': 500, 'array_name_data_a': 'Data_a_0002', 'sample_count_i': 15000, 'stream_number_i': 1, 'time_table_n_i': 0, 'event_number_i': 2, 'array_name_SOH_a': '', 'response_table_n_i': 0, 'array_name_log_a': '', 'sample_rate_multiplier_i': 1, 'time/epoch_l': 1502294430, 'time/type_s': 'BOTH', 'raw_file_name_s': '3ch.fcnt', 'receiver_table_n_i': 1, 'array_name_event_a': '', 'time/ascii_s': 'Wed Aug  9 16:00:30 2017', 'channel_number_i': 1}
```
This seems to be consistent with the documentation. https://github.com/PIC-IRIS/PH5/wiki/Receivers-Group

Is there a different way we could think about this test?

damhuonglan replied:
@timronan you are correct that das tables don't have SEED info. You have to define SEED info in array tables. The info that links DAS tables and array table includes: 
+ Array: das_serial_number_i, channel_number_i, sample_rate_i, sample_rate_multiplier_i, deploy and pickup time
+ Das: channel_number_i, sample_rate_i, sample_rate_multiplier_i, time and sample_count_i

timronan replied:
Based on testing:
```
    Array: das_serial_number_i, channel_number_i, sample_rate_i, sample_rate_multiplier_i, deploy and pickup time
    Das: channel_number_i, sample_rate_i, sample_rate_multiplier_i, time and sample_count_i

```
All seem to be non-unique identifiers, there are multiple DAS tables returned for each Array:das_serial_number_i. Each one of these DAS tables may or may not have matching channel number, sample rate identifiers, sample_rates, time and sample_count_i correspounding to deploy and pickup times. Different sample rates can lead to different count numbers and different Endtimes.  

This pull request is based on the issue 424 which describes that the DAS and Array tables have inconsistent sample rates so no data is returned. 

We could make a test similar to the test_ph5toms.

```
   def test_create_trace(self):

        ph5API_object = ph5api.PH5(
            path=os.path.join(os.getcwd(), 'ph5/test_data/ph5'),
            nickname='master.ph5')
        ph5toms = PH5toMSeed(ph5API_object)
        ph5toms.process_all()
        cuts = ph5toms.create_cut_list()
        for cut in cuts:
            trace = ph5toms.create_trace(cut)
            if trace is not None:
                self.assertEqual(trace[0].stats.sampling_rate, cut.sample_rate)
```

Also, the PH5Availabilty patch was previously tested on experiment 04-24 which is how the issue was originally discovered. The metadata for experiment 04-24 has been changed. 

The original test_ph5availability.py code provides test coverage for this patch. The two arrays refer to the same DAS serial number but have different sample rates. That is exactly what this pull request is trying to isolate and fix. 

```
/Experiment_g/Sorts_g/Array_t_002
	id_s=0407
	deploy_time/epoch_l=1545085230
	deploy_time/micro_seconds_i=0
	pickup_time/epoch_l=1547489524
	pickup_time/micro_seconds_i=0
	das/serial_number_s=5553
	seed_band_code_s=H
	sample_rate_i=200
	sample_rate_multiplier_i=1
	seed_instrument_code_s=H
	seed_orientation_code_s=N

/Experiment_g/Sorts_g/Array_t_003
	id_s=0407
	deploy_time/epoch_l=1545085230
	deploy_time/micro_seconds_i=0
	pickup_time/epoch_l=1547489524
	pickup_time/micro_seconds_i=0
	das/serial_number_s=5553
	seed_band_code_s=L
	sample_rate_i=100
	sample_rate_multiplier_i=1
	seed_instrument_code_s=H
	seed_orientation_code_s=N
```
Does this seems satisfactory?

damhuonglan replied:
It is supposed to be unique group identifier for the correct one. The following is how to identify which time slots in das table match with a station entry in array table.
+ From Array table's station entry, you identify with das table to look at by das serial number. 
+ Then match channel_number_i and sample rate with the time slots in das table. 
+ Check if the time slot (time and sample count) is in the time range defined by station entry's deploy time and pickup time. One station entry may have multi das's time slots. If a station entry have no das tabel's time slot for it or a time slot not belong to a station entry, it is suppose to be report when running ph5validate.

Go back to the data needed to test your change, you need to create array table with 2 station entries with same SEED channel info, same channel_number_i but different deploy/pickup time and sample rate. You also need das data that match with those entries. Those are the kind of data the old code cannot pick up before and your code improve that. Because all the current tests are passed by old code, we need to point out what differences the changes make.

timronan replied:
The problem is that there can be multiple DAS tables with the same serial number that can or cannot have consistent sample rates, sample counts,  start_epoch and stop_epoch. To get DAS tables using array table information one must use the function ph5api.query_das_t, which only requires DAS serial number as an input argument. If start_epoch and/or stop_epoch are not supplied as arguments to ph5api.query_das_t, then the function looks for matching DAS serial numbers through all of time. Unless we want to limit the functionality of ph5api.query_das_t, and likely refactor some code, we cannot make a time based solution. Even if we changed ph5api.query_das_t to require start_epoch and stop_epochs than the solution could still be none unique because PH5 allows DAS tables with consistent serial numbers to have inconsistent sample rates and the check described above could fail. 

I added a test that is an exact match of the failing cases in the data repository. Refer to:
PH5/ph5/test_data/samplerate/
The new test cases are:
test_ph5_availability.test_check_samplerate
test_ph5toms.test_mismatch_sample_rate

I updated ph5toms and ph5availabilty to compare the sample rates between the DAS and array tables, if they match we use this sample rate and proceed. 

If the array tables sample rate does not match any of the returned DAS table sample rates, than we compare all of the returned to DAS table sample rates to themselves. If all of the DAS tables have the same sample rate we assume the metadata is incorrect  and use the DAS table sample rate.

If there are inconsistent sample rates in the DAS tables and they don't match the Array table sapmle rates than we throw an error and the metadata or data must be manually corrected by the PI. This is the logic:

```
                        Das_t = self.ph5.query_das_t(
                            ph5_das,
                            chan=ph5_channum,
                            start_epoch=ph5_start_epoch,
                            stop_epoch=ph5_stop_epoch)
                        for das in Das_t:
                            # Does Array.sr == DAS.sr? If so use sr
                            if das['sample_rate_i'] == st['sample_rate_i']:
                                samplerate_return = das['sample_rate_i']
                                ph5_sr = das['sample_rate_i']
                                times = self.ph5.get_availability(ph5_das,
                                                                  ph5_sr,
                                                                  ph5_channum,
                                                                  starttime,
                                                                  endtime)
                                empty_times = False
                            else:
                                continue
                        if empty_times is True:
                            for i, das in enumerate(Das_t):
                                # IF DAS.SR != Array.SR, USe DAS.SR if match
                                # Checks to see if all DAS tables have same SR
                                sr_prev = Das_t[i-1]['sample_rate_i']
                                if das['sample_rate_i'] != sr_prev:
                                    sr_mismatch = True
                            if sr_mismatch is True:
                                # Else throw warning and fail
                                LOGGER.error('DAS and Array Table sample' +
                                             ' rates do not match, DAS table' +
                                             ' sample rates do not match.' +
                                             ' Data must be updated.')
                            else:
                                # Uses SR if consistent
                                samplerate_return = das['sample_rate_i']
                                ph5_sr = das['sample_rate_i']
                                times = self.ph5.get_availability(ph5_das,
                                                                  ph5_sr,
                                                                  ph5_channum,
                                                                  starttime,
                                                                  endtime)
```
This logic passes all test cases. Master will fail the test cases:

test_ph5_availability.test_check_samplerate
test_ph5toms.test_mismatch_sample_rate

timronan commented:
Has anyone had time to review the additional test and code that was requested in this PR?

damhuonglan commented:
> Has anyone had time to review the additional test and code that was requested in this PR?

Will do it tomorrow. I am waiting until you push all commits needed.

timronan commented:
Sounds good, please refer to my comments above when reviewing. Thanks for continuing to look into this.

damhuonglan commented on ph5/clients/ph5availability.py:
```diff
@@ -473,22 +521,70 @@ def get_availability(self, station='*', location='*',
+                                             ' Data must be updated.')
+                            else:
+                                # Uses SR if consistent
+                                samplerate_return = das['sample_rate_i']
```
I tried to run the command ph5availability, it created error which is at this line when I removed try ... except: `UnboundLocalError: local variable 'das' referenced before assignment`.

Please consider test for main for this command and ph5toms as well.

timronan replied:
I fixed this error by adding try except blocks as suggested. 

```
                            try:
                                if sr_mismatch is True:
                                    LOGGER.error('DAS and Array Table sample' +
                                                 ' rates do not match, DAS' +
                                                 ' table sample rates do not' +
                                                 ' match. Data must be'
                                                 ' updated.')
                                else:
                                    # Uses SR if consistent
                                    samplerate_return = das['sample_rate_i']
                                    psr = das['sample_rate_i']
                                    early, end = self.ph5.get_extent(ph5das,
                                                                     chanum,
                                                                     psr,
                                                                     starttime,
                                                                     endtime
                                                                     )
                            except(UnboundLocalError):
                                continue
```

This, or similar, was included in PH5/ph5/clients/ph5availability/get_availability_extent() and PH5/ph5/clients/ph5availability/get_availability(), PH5/ph5/clients/ph5toms/create_trace()

damhuonglan commented on ph5/clients/ph5toms.py:
```diff
@@ -508,44 +508,66 @@ def create_trace(self, station_to_cut, mp=False):
                                       chan=stc.component,
                                       sample_rate=actual_sample_rate,
                                       apply_time_correction=nt, das_t=das)
+
```
Please use `git diff` to compare so you can avoid unnecessary changes. Last time ph5api.cut() only has a little change but with the extra blank line adding/ removing everywhere in the file, it looked like it had been changed a lot.

timronan replied:
These are flake8 corrections. I somehow would miss them. I wasn't catching them on my local system but I will try and be a little more careful.

damhuonglan reviewed (commented):
I didn't go deep to the logic of your code but see that your code changed since last time I reviewed it. Please make sure it works the way you want. I recommend you to request Data Group @ascire-pic to test ph5toms when you are sure you don't change any code's logic anymore. Please have a checklist so they are sure what to check. About ph5_availability I think only DMC use it, datagroup will not know how to check it.

If you look at our existing test data you can see that we don't add ph5 data in that. If you are able to to add raw data and create ph5 in the test like in [ph5/utilities/tests/test_segd2ph5.py line 36](https://github.com/PIC-IRIS/PH5/blob/38981efab481935f8d7596a6e61df459083763ac/ph5/utilities/tests/test_segd2ph5.py#L36-L38) that would be better. But if that is impossible, we may consider leaving the data like that. You can even use the existing data to create ph5 and tweak it using keftoph5 to create the ph5 data you want.

timronan commented on ph5/clients/tests/test_ph5toms.py:
```diff
@@ -203,6 +205,32 @@ def test_get_nonrestricted_segments(self):
+            if trace is not None:
+                self.assertEqual(trace[0].stats.sampling_rate, cut.sample_rate)
+
+    def test_mismatch_sample_rate(self):
+        self.ph5test_srpath = os.path.join(os.getcwd(),
+                                           'ph5/test_data/samplerate')
+        self.ph5_sr = ph5api.PH5(path=self.ph5test_srpath,
+                                 nickname='master.ph5')
+        ph5toms = PH5toMSeed(self.ph5_sr)
+        ph5toms.process_all()
+        cuts = ph5toms.create_cut_list()
+        for cut in cuts:
+            trace = ph5toms.create_trace(cut)
+            if trace is not None:
+                self.assertEqual(trace[0].stats.station, '10075')
+
```
Adding a test that captures this error message is possible if I add another master.ph5 file. I already have created this master.ph5 file and have created a test for it, but it is mentioned that master.ph5 files are not desirable in the repository. I am wondering if it is worth adding 13 MB for test coverage for a specific log message? I cannot find an obvious way to update the DAS table in the original master.ph5 after it is read. I have found multiple ways to get information from the das tables including `self.ph5_API.ph5.get_node('/Experiment_g/Receivers_g/Das_g_9297/Das_t')` and a loop around the `ph5_API.query_das_t` function, but there do esnot seem to be any obvious setter functions for the DAS table (Please correct me if I am wrong).

I changed the test cases to use the setUp() function. See 
PH5/ph5/clients/tests/test_samplerate.py. 

```    def setUp(self):
        super(TestPH5AvailabilitySampleRate, self).setUp()
        self.ph5test_path_sr = os.path.join(self.home,
                                            'ph5/test_data/samplerate')
        self.ph5_sr = ph5api.PH5(path=self.ph5test_path_sr,
                                 nickname='master.ph5')
        self.sr_avail = ph5availability.PH5Availability(self.ph5_sr)
```

timronan commented on ph5/clients/tests/test_ph5availability.py:
```diff
@@ -1812,6 +1809,19 @@ def test_get_report(self):
+                                                    endtime=None,
+                                                    include_sample_rate=False)
+        self.assertEqual(6, len(ret))
+
```
I changed the test cases to use the setUp() function. See 
PH5/ph5/clients/tests/test_samplerate.py and above. 

I added a test for get_extent. 
```
        ret2 = self.sr_avail.get_availability_extent(station='10075',
                                                     channel='*',
                                                     starttime=None,
                                                     endtime=None,
                                                     include_sample_rate=True)
        self.assertEqual(3, len(ret2))
```

timronan commented:
I built a compiled ph5 data set that is broken in a specific way that fails in the master branch but passes with the fix_424 patch.  It is a lot of additional development work to redesign and rebuild a test . It seems like compiling a PH5 data set from SEGD, or another file type, is not an isolated test case and would test other portions of the PH5 code stack. This pull request is specifically trying to deal with handling instances when the DAS and Array tables have inconsistent sample rates. I am afraid testing that more portions of the code will lead to a more convoluted PR.  

@ascire-pic can you please check ph5toms? You can check it by running ph5toms on the master file that is created in the ph5/test_data/ph5 directory, or any other ph5 experiment where the sample rates match in the Array and DAS tables. In cases with consistent sample rates the output for ph5toms should match between the master and fix_424 branches.

After you verify that this patch does not affect the existing logic, please run ph5toms on the master.ph5 file that I provide in ph5/test_data/samplerate. This test case contains a mismatch in the sample rates between the DAS and Array tables, but all of the sample rates in the DAS tables are consistent. This example is a compressed version of experiment 13-012, which is currently in the DMC archives. If you start on the master branch ph5toms will not return data, for it will not be able to find the DAS tables. Switch to the fix_424 branch and run ph5toms and ms data will be returned. This is a case that we are trying isolate and fix. We need to make sure that all of the data that is stored in the DMC is deliverable, which currently it is not. Thanks for the feedback on this pull request.

damhuonglan commented:
@timronan I see that you got difficult on creating data for testing. Do you want me to check if I can help with that? Then if it is too hard we can consider accepting the ph5 data that you already created.
Let me know if you need help.

timronan commented:
@damhuonglan I think that I figured out the bones of it, the master files are built in:
`https://github.com/PIC-IRIS/PH5/blob/master/create_ph5.sh`.

I think I see how create_ph5.sh works and can try to reconfigure the test suite soon. Can I contact you directly over email if I have any questions? I feel like we may run into some other PH5 problems in the future that could be good to talk about off line. My email is tronan@iris.washington.edu.

timronan commented:
@damhuonglan I decomposed my master.ph5 file and mini.ph5 files into individual components, the test Sample Rate PH5 is recompiled similar to the other test in the PH5 test suite. I also added test for checking messaging and added warning messages when the DAS sample rate is used rather than the Array sample rate. This additional warning message will allows us to track down PH5 datasets with mismatched sample rates so they can be fixed. This will also help in the DMC's validation process. Please review the updated PR when you have time. I have reviewed PH5toMS and PH5Availability and they work as I would expect them to.

damhuonglan commented on create_ph5.sh:
```diff
@@ -20,12 +20,24 @@ sort_kef_gen -n master.ph5 -a > ../metadata/sort_t.kef
+keftoph5 -n master_samplerate.ph5 -k ../../samplerate/all_arrays.kef -p ../samplerate
+keftoph5 -n master_samplerate.ph5 -k ../../samplerate/Expirement_SampleRate.kef -p ../samplerate
+keftoph5 -n master_samplerate.ph5 -k ../../samplerate/Receiver_SampleRate.kef -p ../samplerate
+mstoph5 -n master_samplerate.ph5 -f ../../samplerate/8H.10075..GH1.2012-08-27T23.01.00.ms
+mstoph5 -n master_samplerate.ph5 -f ../../samplerate/8H.10075..GH1.2012-08-31T17.52.00.ms
+mstoph5 -n master_samplerate.ph5 -f ../../samplerate/8H.10075..GH2.2012-08-27T23.01.00.ms
+mstoph5 -n master_samplerate.ph5 -f ../../samplerate/8H.10075..GH2.2012-08-31T17.52.00.ms
+mstoph5 -n master_samplerate.ph5 -f ../../samplerate/8H.10075..GHZ.2012-08-27T23.01.00.ms
+mstoph5 -n master_samplerate.ph5 -f ../../samplerate/8H.10075..GHZ.2012-08-31T17.52.00.ms
```
Use option -d for all miniseed files in the directory

damhuonglan commented on create_ph5.sh:
```diff
@@ -20,12 +20,24 @@ sort_kef_gen -n master.ph5 -a > ../metadata/sort_t.kef
+keftoph5 -n master_samplerate.ph5 -k ../../samplerate/Das_SampleRate.kef
+mkdir error
+cd error
+keftoph5 -n master_samplerate_error.ph5 -k ../../../samplerate/all_arrays.kef -p ../error
+keftoph5 -n master_samplerate_error.ph5 -k ../../../samplerate/Expirement_SampleRate.kef -p ../error
+keftoph5 -n master_samplerate_error.ph5 -k ../../../samplerate/Receiver_SampleRate.kef -p ../error
```
Can copy master file from samplerate/ before miniseed info is added into it.

damhuonglan commented on create_ph5.sh:
```diff
@@ -20,12 +20,24 @@ sort_kef_gen -n master.ph5 -a > ../metadata/sort_t.kef
+keftoph5 -n master_samplerate_error.ph5 -k ../../../samplerate/all_arrays.kef -p ../error
+keftoph5 -n master_samplerate_error.ph5 -k ../../../samplerate/Expirement_SampleRate.kef -p ../error
+keftoph5 -n master_samplerate_error.ph5 -k ../../../samplerate/Receiver_SampleRate.kef -p ../error
+mstoph5 -n master_samplerate_error.ph5 -f ../../../samplerate/8H.10075..GH1.2012-08-27T23.01.00.ms
```
After merge with master, change option -f to -r

damhuonglan commented on create_ph5.sh:
```diff
@@ -20,12 +20,24 @@ sort_kef_gen -n master.ph5 -a > ../metadata/sort_t.kef
+keftoph5 -n master_samplerate_error.ph5 -k ../../../samplerate/Receiver_SampleRate.kef -p ../error
+mstoph5 -n master_samplerate_error.ph5 -f ../../../samplerate/8H.10075..GH1.2012-08-27T23.01.00.ms
+keftoph5 -n master_samplerate_error.ph5 -k ../../../samplerate/Das_SampleRate_error.kef
+cd ../../
```
Not necessary.

damhuonglan commented on ph5/clients/tests/test_ph5toms.py:
```diff
@@ -5,6 +5,9 @@
 import unittest
 from ph5.clients.ph5toms import StationCut, PH5toMSeed
 import copy
+from ph5.core import ph5api
+import os
+from testfixtures import LogCapture
```
Reorganize import: stdlib packages/ skip line/ other packages/ skip line/ project's packages

damhuonglan commented on ph5/clients/tests/test_ph5toms.py:
```diff
@@ -203,6 +206,35 @@ def test_get_nonrestricted_segments(self):
             station_to_cut_list, restricted_list)
         self.assertEqual(result[0].__dict__, expected.__dict__)
 
+    def test_create_trace(self):
```
Separate test_create_trace() and test_mismatch_sample_rate to a different class: TestPH5toMSeed_samplerate(LogTestCase, TempDirTestCase) to use tearDown to delete self.ph5_object (use self.ph5_object instead of ph5API_object, self.ph5_sr to utilize tearDown())
LogTestCase help prevent log information to be printed to terminal in unittest.
TempDirTestCase create self.home is the root directory of the project, self.tmpdir is the temporary directory that is supposed to keep all files created during unittest. self.tmpdir will be remove with all in it when each test is finished if the test inherit tearDown from this class. The current directory will be self.tmpdir at the beginning of the test..

damhuonglan commented on ph5/clients/tests/test_ph5toms.py:
```diff
@@ -203,6 +206,35 @@ def test_get_nonrestricted_segments(self):
 
+    def test_create_trace(self):
+
+        ph5API_object = ph5api.PH5(
+            path=os.path.join(os.getcwd(), 'ph5/test_data/ph5'),
+            nickname='master.ph5')
```
Change to self.ph5_object to be cleaned up in tearDown()

damhuonglan commented on ph5/clients/tests/test_ph5toms.py:
```diff
@@ -203,6 +206,35 @@ def test_get_nonrestricted_segments(self):
+    def test_mismatch_sample_rate(self):
+        self.ph5test_srpath = os.path.join(os.getcwd(),
+                                           'ph5/test_data/ph5/samplerate')
+        self.ph5_sr = ph5api.PH5(path=self.ph5test_srpath,
+                                 nickname='master_samplerate.ph5')
```
Change to self.ph5_object to be cleaned up in tearDown()

damhuonglan commented on ph5/clients/tests/test_ph5toms.py:
```diff
@@ -203,6 +206,35 @@ def test_get_nonrestricted_segments(self):
+                trace = ph5toms.create_trace(cut)
+                if trace is not None:
+                    self.assertEqual(trace[0].stats.station, '10075')
+        self.assertIsNotNone(log)
```
Using log should be moved inside with.

damhuonglan commented on ph5/clients/tests/test_ph5toms.py:
```diff
@@ -203,6 +206,35 @@ def test_get_nonrestricted_segments(self):
+                if trace is not None:
+                    self.assertEqual(trace[0].stats.station, '10075')
+        self.assertIsNotNone(log)
+        self.ph5_sr.close()
```
Should close in tearDown().

damhuonglan commented on ph5/clients/tests/test_samplerate.py:
```diff
@@ -0,0 +1,85 @@
+"""
```
Test ph5availability methods should be moved into test_ph5availability.py

damhuonglan commented on ph5/clients/tests/test_samplerate.py:
```diff
@@ -0,0 +1,85 @@
+                                 nickname='master_samplerate.ph5')
+        self.sr_avail = ph5availability.PH5Availability(self.ph5_sr)
+
+    def tearDown_sr(self):
```
Adding '_sr' to tearDown makes it not to be called => change to tearDown

damhuonglan commented on ph5/clients/tests/test_samplerate.py:
```diff
@@ -0,0 +1,85 @@
+            trace = ph5toms.create_trace(cut)
+            if trace is not None:
+                self.assertEqual(trace[0].stats.station, '10075')
+        self.ph5_sr.close()
```
Clean up in tearDown()

damhuonglan commented on ph5/clients/tests/test_samplerate.py:
```diff
@@ -0,0 +1,85 @@
+                                                 endtime=None,
+                                                 include_sample_rate=True)
+        self.assertEqual(6, len(ret))
+        self.assertIsNotNone(log)
```
Move inside with

damhuonglan commented on ph5/clients/tests/test_samplerate.py:
```diff
@@ -0,0 +1,85 @@
+                                                         starttime=None,
+                                                         endtime=None)
+        self.assertIsNotNone(log2)
+        self.ph5_sr.close()
```
Clean up in tearDown

damhuonglan commented on ph5/clients/tests/test_samplerate.py:
```diff
@@ -0,0 +1,85 @@
+        self.assertIsNotNone(log2)
+        self.ph5_sr.close()
+
+    def test_availability_error(self):
```
Move to a different class because it doesn't use any info from setUp of the current class.

damhuonglan commented on ph5/clients/tests/test_samplerate.py:
```diff
@@ -0,0 +1,85 @@
+                                              starttime=None,
+                                              endtime=None,
+                                              include_sample_rate=True)
+        self.assertEqual(log_error[2][2], 'DAS and Array Table sample rates do'
```
Move inside with.
I don't know how it work here. When I run this test in my computer, it creates error because log_error isn't iterative. May use log_error.records[2].msg

timronan replied:
My local environment is different and this updated test does not work on my enviroment, but it is fine to use whatever index works on @damhuonglan's enviroment.

damhuonglan reviewed (commented):
It looks good now. Just need to make it to follow standard code type as I suggested here. Actually I have done that myself in #438 . If you don't understand my comments, you can refer to that branch.

damhuonglan commented on ph5/clients/ph5toms.py:
```diff
@@ -508,44 +508,73 @@ def create_trace(self, station_to_cut, mp=False):
+                            actual_sample_rate = 0
+                except(UnboundLocalError):
+                    continue
+            if actual_sample_rate != 0:
```
I got error "local variable 'actual_sample_rate' referenced before assignment" when run this for the master file inside ph5/test_data/ph5/samplerate/error.

timronan replied:
Fixed in fix_424_help. All that need to be done was add return after log.error messages.

timronan commented:
If all of these reformatting suggestions, none of them seem to be logical changes, have been completed on branch fix_424_help than we should review pull request #438 and merge if it is satisfactory. We can close this PR once fix_424_help  is completed.

damhuonglan commented:
> If all of these reformatting suggestions, none of them seem to be logical changes, have been completed on branch fix_424_help than we should review pull request #438 and merge if it is satisfactory. We can close this PR once fix_424_help is completed.
I don't think it's a good way. I think you can merge fix_424_help to fix_424 then merge fix_424 to master in this PR.

ascire-pic commented:
@akram-pic is going to test this.

akram-pic commented:
As requested, I have tested fix_424 (ph5toms) to see if the changes in fix_424 had affected the functionality of ph5toms.
I have chosen two datasets that have data in multiple sample rates (16-015_Wavefields and 18-027_BFNO). 
It appears that several options including -r, --channel, and -a do not work anymore when they are used simultaneously with -s, -t (start-end). 

Examples:

These command works in ph5toms version 2019.93 currently on ph52

*** ph5toms -n master.ph5 -a 007 --station 516 -s 2016-07-20T14:35:14.010000Z -t 2016-07-20T14:40:14.010000Z -F MSEED -o mseed-test-fix_424/before-fix/

***ph5toms -n master.ph5 --channel LHZ,HHZ --station 516 -s 2016-07-20T14:35:14.010000Z -t 2016-07-20T14:40:14.010000Z -F MSEED -o mseed-test-fix_424/before-fix/

***ph5toms -n master.ph5 -r 100,1 --station 516 -s 2016-07-20T14:35:14.010000Z -t 2016-07-20T14:40:14.010000Z -F MSEED -o mseed-test-fix_424/before-fix/

The same exact commands above do not work with fix_424. Only when I remove -s and -t options, then they will work.
All statements above applies to both format SEED and SAC.

damhuonglan commented:
> The same exact commands above do not work with fix_424. Only when I remove -s and -t options, then they will work.
@akram-pic Can you describe in more details please? If the result you got is different, how different is it? If you got any error, can you add the error in a comment please?

akram-pic commented:
> > The same exact commands above do not work with fix_424. Only when I remove -s and -t options, then they will work.
> 
> @akram-pic Can you describe in more details please? If the result you got is different, how different is it? If you got any error, can you add the error in a comment please?

@damhuonglan  
I did not get any error message. Running the commands under the new fix with -s and -t option simply did not result in anything and it did not throw any error message. Running the command would just quickly end and I would have my prompt back as if it was working, but my output directory would be empty. 

You can test this here:

(ph5) ph51@ph5-1:/external/ph5/Completed/16-015_Wavefields$ ph5toms -n master.ph5 -r 1 --station 516 -s 2016-07-20T14:35:14.010000Z -t 2016-07-20T14:40:14.010000Z -F MSEED -o mseed-test-fix_424/after-fix/

timronan commented:
This is happening because no das tables are returned when ph5toms calls ph5api.query_das for your provided examples.  

There are two logical patterns in ph5api.query_das_t, one that uses a sampling rate and one that does not, both of these logical patterns were in place before fix_424, check lines 803 and 835 in https://github.com/PIC-IRIS/PH5/blob/master/ph5/core/ph5api.py. Fix_424 has not changed any of this logic. 

These are the logical checks:

line 803-809
```
            for row in tbl.where(
                    '(channel_number_i == '
                    + str(chan) + ' )&(epoch_i+micro_seconds_i/1000000>='
                    + str(start_epoch) +
                    ')&(epoch_i+micro_seconds_i/1000000<='
                    + str(stop_epoch) + ')'
            ):
```
lines 835-845:
```
            for row in tbl.where(
                    '(channel_number_i == '
                    + str(chan) + ' )&(epoch_i+micro_seconds_i/1000000>='
                    + str(start_epoch) +
                    '-sample_count_i/sample_rate_i/sample_rate_multiplier_i)'
                    '&(epoch_i+micro_seconds_i/1000000<='
                    + str(stop_epoch) + ')&(sample_rate_i==' +
                    str(sample_rate) +
                    ')&(sample_rate_multiplier_i==' +
                    str(sample_rate_multiplier) + ')'
            ):
```

The logic is working inconsistently with these checks. In the provided examples the master branch uses the line 835 check which extends the input starttime `s` by `-sample_count_i/sample_rate_i/sample_rate_multiplier_i seconds` 

`(epoch_i+micro_seconds_i/1000000>=+ str(start_epoch) +'-sample_count_i/sample_rate_i/sample_rate_multiplier_i)'`

This seems like a bug and should maybe be removed from the master branch. It doesn't make sense to automatically shift a user's startdate by all of the samples in the trace. Why would the user provide the startdate if we just shift it? I am also a little skeptical about the output miniseed data from this shifted query. Is the miniseed file's timing accurate to the DAS table? I have not checked and do not know.

If this line of code, master branch ph5api 839,  is removed, and the logic between the line 803 and line 835 checks become consistent, than the Akram's example will work consistently. In both cases no data will be returned because the DAS table's epoch is less than the requested epoch. I am not sure how to move forward, so let me know what to do. This is a good catch.

damhuonglan commented:
> This seems like a bug and should maybe be removed from the master branch.

The one in ph52 is the latest update of master branch. Master branch has been merged to fix_424 before the test. So if it is a bug that has been fixed in master branch it should be fixed here too. I will spend time this morning to figure out where the problem is.

damhuonglan commented:
@timronan in ph5toms.py, you removed sample_rate and sample_rate_multiplier when calling query_das_t():
https://github.com/PIC-IRIS/PH5/blob/73c211a23a3dbc0869a3d56936eec547bf6be5b7/ph5/clients/ph5toms.py#L514-L516
It results that query_das_t() has to use default sample_rate=None which lead to the following condition is satified:
https://github.com/PIC-IRIS/PH5/blob/73c211a23a3dbc0869a3d56936eec547bf6be5b7/ph5/core/ph5api.py#L801

timronan commented:
Yeah, that is correct. That is what I was outlining earlier. Look at the logic for how the if and else statements search for rows.
The `if` statement does  a logical check just on the  input start time. `startepoch` is the user supplied input. 
`epoch_i+micro_seconds_i/1000000>= str(start_epoch) `

The `else` statement extends the start time start_time  by the length of the trace. 
`epoch_i+micro_seconds_i/1000000>=str(start_epoch)-sample_count_i/sample_rate_i/sample_rate_multiplier_i`

The master branch uses the else statement fix_424 uses the if statement. The question is: why is the user included starttime extended by the trace? Is `epoch_i` the time stamp for the last sample in the trace?

timronan commented:
In the `ph5toms -n master.ph5 --channel LHZ,HHZ --station 516 -s 2016-07-20T14:35:14.010000Z -t 2016-07-20T14:40:14.010000Z -F MSEED -o mseed-test-fix_424/before-fix/` example the master branch is picking up a trace with an epoch_l at 1466359810 which is Sunday, June 19, 2016 6:10:10 PM. It is then relabeling this trace to YW.516.01.HHZ.2016-07-20T14.35.14.ms, which is Sunday, June 19, 2016 02:35:14 PM.  That seems incorrect. I am going to push the consist logic to fix_424. It breaks some historical test, but I am not going to push the test updates because @PIC-IRIS/ph5-admin should verify this and be the ones to update historical test.

damhuonglan commented:
> In the `ph5toms -n master.ph5 --channel LHZ,HHZ --station 516 -s 2016-07-20T14:35:14.010000Z -t 2016-07-20T14:40:14.010000Z -F MSEED -o mseed-test-fix_424/before-fix/` example the master branch is picking up a trace with an epoch_l at 1466359810 which is Sunday, June 19, 2016 6:10:10 PM. It is then relabeling this trace to YW.516.01.HHZ.2016-07-20T14.35.14.ms, which is Sunday, June 19, 2016 02:35:14 PM. That seems incorrect. I am going to push the consist logic to fix_424. It breaks some historical test, but I am not going to push the test updates because @PIC-IRIS/ph5-admin should verify this and be the ones to update historical test.

I ask DG to check the files created in Akram's test. They use msi to test and it looks correct to them. I tested it myself:
```
$ msi YW.516..LH1.2016-07-20T14.35.15.ms
YW_516__LH1, 000001, D, 4096, 300 samples, 1 Hz, 2016,202,14:35:14.640000
```
The data in ms file match with the name.

damhuonglan commented:
This is the issue and PR when they added `-sample_count_i/sample_rate_i/sample_rate_multiplier_i` to the condition.
https://github.com/PIC-IRIS/PH5/issues/255
https://github.com/PIC-IRIS/PH5/commit/46ba629b17271f080e33e794e8c1b0b732b15efa#diff-570207f107c1035bad4288092c63ab23df0bc00e039694dc76295b20f820ae14
It isn't a bug. It is to fix a bug.

timronan commented:
> It isn't a bug. It is to fix a bug.

Fair enough, thanks for checking this. I updated the logic of `query_das_t ` to extend the starttime for ph5toms and ph5availabiity the same way that master works. I used an elif statement: 

```
        elif check_samplerate is False:
            for row in tbl.where(
                 '(channel_number_i == '
                 + str(chan) + ' )&(epoch_i+micro_seconds_i/1000000 >= '
                 + str(start_epoch) +
                 '-sample_count_i/sample_rate_i/sample_rate_multiplier_i)'
                 '&(epoch_i+micro_seconds_i/1000000 <= '
                 + str(stop_epoch) + ')'
                 ):
```
See commit 613a60f3f560005575ad9d1e7324f4c172ce8bea. This logic seems to pass all of @akram-pic's test cases.

akram-pic reviewed (approved):
I have tested fix_424 (mstoph5) with as many options as possible on the Wavefields dataset  (-l, -r, -c, -t, -s, --channel, -a, -station, -d, -F), and I inspected the outputs with msi.  All combinations resulted in outputs, and all output files were as expected and correct as far as I could say.

timronan commented:
@akram-pic this is great, thanks for testing this code. Are there any other steps that should be taken before this branch is merged into the Master branch? Who should perform the merge?

damhuonglan commented:
> @akram-pic Are there any other steps that should be taken before this branch is merged into the Master branch? Who should perform the merge?

I think it's good to be merged and you can do it yourself.
